### DISCUSSION TITLE
Using more substitutions to outsource config settings for easier update aso.

### DISCUSSION THREAD

**Discussion by Lin4fun** [BigBobbas/ESP32-S3-Box3-Custom-ESPHome#39] [Ideas]
Hi,

Just came over that it is painfull to do an update if something has changed in the configs and an update is needed.
So it could be usefull to use the substitutions and variables for media player and audio output as well simple light configuration and all other entitities. Same for OTA, encryption,ap,wifi and so on.

Doing so it is much easier to do an update. The "base config" could hold those informations and the functional yaml and all other could be loaded from github during compile.

The config is heavy and could be to much for the most beginners. When I started to use this yaml it took some time until i understood everything in the config.

The most perfect solution of course would be if the hardware could be configured using drop down at the device config. But I have no glue if this is even possible yet in HA/esphome devices / esphome dashboard or through hacs some how.

**Comment by BigBobbas** [comment]
Hi , 
thanks for your feedback and suggestions, others ideas and views are always welcome. Rather than me just doing what **I** think would be good.

I will be looking into using substitutions for certain things and hopefully the addition of entities. 
With regards to having a separate base config that is also something that I am considering.  Part of the reason that I started this project other than to provide users with a way to get the most out of the S3box3 was to also give users the ability to understand how the config is constructed, so that they can also learn what is needed to work with a yaml config and to make changes to 'take control' in the same way that I have had to. 

Although the config, as you say is rather heavy. I do like that users are able to fairly easily follow the guide to make their own changes and can see the whole config in one place. Having the config as a one click install or with includes/packages removes the ability for the user to get a full view of how things are constructed. However I do realise that for some this is very daunting and they may like the easier route of click and install , then choose their entities and away they go. So i'm hoping to add both options in the future.

Thanks so much for your input.
Cheers

Rob

**Reply by Lin4fun** [reply]
I fully understand your point and I have to say that this also challanged me to get a deeper understanding how things work. To get controle over the existing yaml I used ChatGPT  by pasting parts of the code and ask for deeper explaination.

But there are as well a lot of people who do not have the understanding to fight theirselves through the code because they do miss a lot of knowledge. 
I think if thoose users want to use the config they can use the webinstaller version with well tested config. This would may also reduce the support. For all other who also would like to contribute through testing for example can use the beta channel which have to be compiled with config.  In the end there is the dev channel, for the people with deep understanding.

By outsourcing to several files it would be easier to understand the code. And everybody who is messing around with code should know what an variable or substitution is. It is just much more structurized. We all know how important structure is for coding to keep controle.

I personally was very happy when I found the work you are doing here and I am very sure that this project will take off! By this the support will raise expotentially and nothing is more bad if you need than to back off because this project is taking over your life.

But that is of course just how I do see the this and if everything will stay like it is I am also very happy. This project is "that thing" i was looking for the past to years since i started with homeassistant / esphome.

How ever, thank you for your work!
